# Slurm script that utilizes EggNOG and a python script to find the completion of KEGG modules
## Info
This script allows for finding KEGG module completion without using the KEGG webserver. It circumvents
the use of blastKOALA by assessing the K terms present in the genome using EggNOG. A python script is used
to parse the KEGG definitions of each module, and to then calculate each possible combination of K terms 
that complete this module into a functional pathway. The script then assess per module which combination 
is most complete given the genome's K terms, and outputs the degree of completion. 

This script is a .slurm script that (after giving your input) can be run directly on the OIST Deigo server 
using "sbatch KEGGstand.slurm"

## Important notes: 
EggNOG appears to give more results than blastKOALA (approx. 3% more K terms). Expect to find modules to be 
slightly more complete than when running the KEGG workflow. I do not know whether these additional hits are 
due to better sensitivity (more true positives), or due to less stringency (more false positives).

For now the python script is still undergoing validation. My tests show it performs well, but until more
exhaustive tests are done, please perform some validation of your results. Until I am more confident of its
performance, the script will also not be pushed to github. (It can be called from Deigo normally, as per the script). 

## Input
The script takes a directory as input, and analyzes each fasta file in the directory. By default it takes genome 
fasta files, but it can be modified to take CDS or protein sequences instead (which will run much faster, although they could
potentially be less sensitive). 

The input is hardcoded, so needs to be adjusted in the script itself.
Please change the following lines to suit your case:

Line 8: #SBATCH --array=x-x <-- Change the array size to fit your number of samples (0-0 = 1 sample, 0-1 = 2 samples etc)

Line 16: input_dir=/path/to/dir <-- Change the path to the path of your directory containing the fasta files

Line 17: output_dir=/path/to/dir <-- Change the path to your desired output folder (probably somewhere on /flash)

Line 18: extension=.fasta <-- Change "fasta" to the extension used in your fasta file (usually fna, faa, fa, or fasta)

If you want to run on CDS or proteins you will have to change line 33 as well:

CDS: --itype genome -> --itype CDS --translate (--translate is optional, but should be faster)

Proteins: --itype genome -> --itype proteins 

If all variables are correct, just run the script using "sbatch KEGGstand.slurm"

## Output
The output will be generated in the specified folder. It includes the regular EggNOG output, and 2 text files generated by 
the python script. By default these are:

input_fasta.emapper.annotations_KEGG_complete_modules.tsv = Tab-delimited text file only containing the modules found to be 
100% complete. First column is the module name, the second one lists the gene K terms found to comprise this complete module. 

input_fasta.emapper.annotations_KEGG_completion.tsv = Tab-delimited text file showing the found completion for each KEGG
module. First column is the module name, the second one shows the completion (as a fraction 0-1), and the third lists the gene 
K terms comprising the most complete possibility for this module. 
